web3_proxy/src/bin/web3_proxy_cli/proxyd.rs

```diff
@@ -143,70 +154,160 @@ async fn run(
+    //     let consensus = spawned_app
+    //         .consensus_connections_watcher
+    //         .borrow_and_update();
+    //
+    //     // Let's just do a super dirty unwrap to get things going
+    //     if consensus.unwrap().backups_needed {
+    //         info!(
+    //             "waiting longer. found consensus with backups: {}",
+    //             consensus.unwrap().head_block.as_ref().unwrap(),
+    //         );
+    //     } else {
+    //         // TODO: also check that we have at least one archive node connected?
+    //         break;
+    //     }
+    // }
     let _ = spawned_app.app.head_block_receiver().changed().await;
 
+
     // start the frontend port
-    let frontend_handle = tokio::spawn(frontend::serve(app_frontend_port, spawned_app.app.clone()));
+    let frontend_handle = tokio::spawn(frontend::serve(
+        app_frontend_port,
+        spawned_app.app.clone(),
+        frontend_shutdown_receiver,
+        frontend_shutdown_complete_sender,
+    ));
+
+    let frontend_handle = flatten_handle(frontend_handle);
 
     // if everything is working, these should all run forever
+    let mut exited_with_err = false;
+    let mut frontend_exited = false;
     tokio::select! {
         x = flatten_handles(spawned_app.app_handles) => {
             match x {
                 Ok(_) => info!("app_handle exited"),
                 Err(e) => {
-                    return Err(e);
+                    error!("app_handle exited: {:#?}", e);
+                    exited_with_err = true;
                 }
             }
         }
-        x = flatten_handle(frontend_handle) => {
+        x = frontend_handle => {
+            frontend_exited = true;
             match x {
                 Ok(_) => info!("frontend exited"),
                 Err(e) => {
-                    return Err(e);
+                    error!("frontend exited: {:#?}", e);
+                    exited_with_err = true;
                 }
             }
         }
         x = flatten_handle(prometheus_handle) => {
             match x {
                 Ok(_) => info!("prometheus exited"),
                 Err(e) => {
-                    return Err(e);
+                    error!("prometheus exited: {:#?}", e);
+                    exited_with_err = true;
                 }
             }
         }
         x = tokio::signal::ctrl_c() => {
+            // TODO: unix terminate signal, too
             match x {
                 Ok(_) => info!("quiting from ctrl-c"),
                 Err(e) => {
-                    return Err(e.into());
+                    // TODO: i don't think this is possible
+                    error!("error quiting from ctrl-c: {:#?}", e);
+                    exited_with_err = true;
                 }
             }
         }
-        x = shutdown_receiver.recv() => {
+        // TODO: how can we properly watch background handles here? this returns None immediatly and the app exits. i think the bug is somewhere else though
```

Comment:
is this todo still true? something is definitely wrong if so.

Reply:
I think this was from one of the merges, so I can't really tell, would be nice if you could maybe check that (didn't touch this file except for the merges really..). I can also see how it plays with the main branch / stats depending on which one is more advanced

Reply:
Yeah I double checked, it's just naturally occuring from the merges. Could you maybe double check this file https://github.com/yenicelik/web3-proxy/blob/stats_v3/web3_proxy/src/bin/web3_proxy_cli/proxyd.rs#L193 and let me know if it seems out of sync with any of the branches? so far I just merged main into stats multiple times, it shouldn't have anything else in there

Reply:
I just double checked, there seem to be a bunch of merge conflicts between `main` and `stats_v2`. Would it be possible if you could perhaps go through them and decide what to keep / discard? I'm not sure if I'm the right person to make these decisions xD

Reply:
basically both branches modified the files on separate commit histories, hence why there are weirdnesses in this file (and more so than else-where)